# Projet_8: NOTEBOOK DE PREPROCESSING IMAGES 

# 1 Introduction

La préparation des données (ou data processing, en anglais) consiste à rassembler, combiner, structurer et organiser les données afin de pouvoir les analyser dans le cadre de programmes d'informatique décisionnelle (BI, Business Intelligence) et d'analytique métier (BA, Business Analytics).

Le travail de préparation est effectué par les équipes des services IT et BI, qui intègrent des ensembles de données pour les charger dans un entrepôt de données, une base de données  ou un référentiel sous forme de lac de données Hadoop

L'un des principaux objectifs de la préparation des données consiste à assurer que les informations concernées sont exactes et cohérentes, afin que les applications BI et BA donnent des résultats pertinents.

Dans les applications de Big Data, la préparation des données est généralement une tâche automatisée.

# 2 Importation des Librairies

In [1]:
import os
import os.path
import shutil

# 3 Localisation du repertoire de developpement de nos scripts sur notre environnement local

In [2]:
cwd=os.getcwd()

In [3]:
cwd

'C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS'

# 4 Création d'un repertoire DataLake de notre environnement local dénommé "Input_Data"

In [4]:
if not os.path.exists("Input_Data"):
    # if the demo_folder directory is not present 
    # then create it.
    os.makedirs("Input_Data")

In [5]:
Path_Input_Data = cwd +'/'+ "Input_Data"

In [6]:
os.listdir( Path_Input_Data )

['Test', 'Training', 'Validation']

# 5 Exploration des données images collectées  dans le repertoire Test fourni par Kaggle

## 5_1 Etape1 : Effectuons une copie du datalakedans notre environnement local repertoire

In [7]:
try:
    shutil.copytree('C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Input_Data','C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Input_Data_Backup')
except FileExistsError:
   # directory already exists
    pass

## 5_2 Etape2 : Effectuons une copie du sous repertoire Test comprise dans notre datalake

In [8]:
try:
    shutil.copytree('C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Input_Data\\Test','C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Test_Backup')
except FileExistsError:
   # directory already exists
    pass

## 5_3 Etape3 : Effectuons une exploration de l'effectif et taille de nos données.

### 5_3_1_Etape3-1: Evaluation du nombre de fichiers images dans le sous repertoire Test

In [9]:
lst_folder=os.listdir("Test_Backup")
print(lst_folder)
len(lst_folder)
print(f"\n Nous avons détécté {len(lst_folder)} catégories de fruits dans notre base de donnée"  )

['apple_6', 'apple_braeburn_1', 'apple_crimson_snow_1', 'apple_golden_1', 'apple_golden_2', 'apple_golden_3', 'apple_granny_smith_1', 'apple_hit_1', 'apple_pink_lady_1', 'apple_red_1', 'apple_red_2', 'apple_red_3', 'apple_red_delicios_1', 'apple_red_yellow_1', 'apple_rotten_1', 'cabbage_white_1', 'carrot_1', 'cucumber_1', 'cucumber_3', 'eggplant_violet_1', 'pear_1', 'pear_3', 'zucchini_1', 'zucchini_dark_1']

 Nous avons détécté 24 catégories de fruits dans notre base de donnée


In [10]:
Count=0
for dirpath, dirs, files in os.walk("Test_backup"):
    Count+=len(files)
print(f"Nous avons détécté {Count} image de fruits dans notre base de donnée Test"  )
print(f"\nNous allons en extraire 10% pour notre projet soit 300 images à peu près" )

Nous avons détécté 3110 image de fruits dans notre base de donnée Test

Nous allons en extraire 10% pour notre projet soit 300 images à peu près


### 5_3_2_ Etape3-2: Evaluation  du nombre de fichiers images

In [11]:
Path_In_Backup = cwd +'\\'+ "Input_Data_Backup"

In [12]:
os.listdir( Path_In_Backup )

['Test', 'Training', 'Validation']

In [13]:
total = 0
for d in os.listdir(Path_In_Backup):
    print(f'\n***{d}****')
    for sub_dir in os.listdir(Path_In_Backup + "/" + d):
        target_sub_dir = Path_In_Backup + '/'+ d +'/'+ sub_dir
        files = os.listdir(target_sub_dir)
        count = len(files)
        total += count
        print(f'{sub_dir}: {count}')
        print(f'Total: {total}')
       
print(f"\n\n Le repertoire Input_Data stocke {total} Images dont les extractions des caractéristiques sur pyspark\
 ne peuvent pas etre      calculées en environnement local.Nous serons obligés de déployer ces données dans un environnement Big Data. ") 

print("\nL'environnement Big Data Selectionné est:AMAZON WEB SERVICE")

print(f"\nNous allons exploiter qu'un effectif de 300 images pour notre environnement de Test soit à peu près un echantillo de 2.5% de l'effectif total testés avant\
deploiement dans un environnement de production.")


***Test****
apple_6: 157
Total: 157
apple_braeburn_1: 160
Total: 317
apple_crimson_snow_1: 159
Total: 476
apple_golden_1: 154
Total: 630
apple_golden_2: 154
Total: 784
apple_golden_3: 158
Total: 942
apple_granny_smith_1: 160
Total: 1102
apple_hit_1: 234
Total: 1336
apple_pink_lady_1: 156
Total: 1492
apple_red_1: 154
Total: 1646
apple_red_2: 159
Total: 1805
apple_red_3: 140
Total: 1945
apple_red_delicios_1: 150
Total: 2095
apple_red_yellow_1: 154
Total: 2249
apple_rotten_1: 159
Total: 2408
cabbage_white_1: 47
Total: 2455
carrot_1: 50
Total: 2505
cucumber_1: 50
Total: 2555
cucumber_3: 81
Total: 2636
eggplant_violet_1: 80
Total: 2716
pear_1: 162
Total: 2878
pear_3: 72
Total: 2950
zucchini_1: 80
Total: 3030
zucchini_dark_1: 80
Total: 3110

***Training****
apple_6: 315
Total: 3425
apple_braeburn_1: 320
Total: 3745
apple_crimson_snow_1: 318
Total: 4063
apple_golden_1: 308
Total: 4371
apple_golden_2: 308
Total: 4679
apple_golden_3: 316
Total: 4995
apple_granny_smith_1: 320
Total: 5315
apple_

### 5_3_3 Etape 3-3 Evaluation de la taille des metadonnées et une contrainte du Service S3 à vérifier :Pas de stokage au dela de 160Go chez AWS Cloud Service sinon négocier une demande spéciale.

In [14]:
def get_size(start_path = '.'):
    total_size = 0
    for dirpath, dirnames, filenames in os.walk(Path_Input_Data):
        for f in filenames:
            fp = os.path.join(dirpath, f)
            # skip if it is symbolic link
            if not os.path.islink(fp):
                total_size += os.path.getsize(fp)

    return total_size

print(get_size(), 'bytes')

584962030 bytes


#### Nous obtenons une taille total de fichier proche de 0.6GByte(0.6Go) ce qui est très en dessous de la contrainte de taille des fichiers à stocker dans le Simple Stockage Service AWS dénommé S3.

### 5_3_4 Evaluation du budget à consacrer à la location des services AWS

##### Nous avons fixé les frais de location des services Cloud Aws à 200euros Max

# 6 Transfert des 12455 images de fruits vers un repertoire unique dénommé Input_Big_Data avec sous repertoires labelisés.

## 6_1 Création du repertoire Input_Big_Data

In [15]:
if not os.path.exists("Input_Big_Data"):
      
    # if the demo_folder directory is not present 
    # then create it.
    os.makedirs("Input_Big_Data")

## 6_2 Script de transfert 

In [16]:
Path_Big_Data= cwd +'/'+ "Input_Big_Data"

for sub_dir_ref in os.listdir(Path_Big_Data + "/" + "Test"):
       for d in os.listdir(Path_Big_Data):
            for sub_dir in os.listdir(Path_Big_Data + "/" + d):
                if  sub_dir==sub_dir_ref:
                        target_sub_dir = Path_Big_Data + '/'+ d +'/'+ sub_dir
                        files = os.listdir(target_sub_dir)
                        for file in files:
                            source = target_sub_dir +'/'+file
                            destination = Path_Big_Data +'/'+"Test" +'/'+sub_dir_ref
                            filename = os.path.basename(source) # In order to overwrite file (if one already exists in destination),
                            dest = os.path.join(destination,filename)#we need to specify full path for destination,
                            shutil.move(source, dest)

                else:           
                       pass
                    
print("Job Done")
os.startfile("Input_Big_Data")                    

Job Done


## 6_3 Verification du transfert des 12455 images vers le datalake local de transit Input_Big_Data/Test.

In [17]:
Count=0
for dirpath, dirs, files in os.walk("Input_Big_Data/Test"):
    Count+=len(files)
print(f"Nous avons détécté {Count} images de fruits dans notre base de données transférées avec label\
   vers le sous repertoire Test rattaché au repertoire Input_Big_Data"  )


Nous avons détécté 12455 images de fruits dans notre base de données transférées avec label   vers le sous repertoire Test rattaché au repertoire Input_Big_Data


## 6_4 Création du repertoire Input_Big_Data

In [18]:
if not os.path.exists("Input_Big_Data/Test1"):
      
    # if the demo_folder directory is not present 
    # then create it.
    os.makedirs("Input_Big_Data/Test1")

## 6_5 Conversion des 12455 images stockées dans le sous repertoire Input_Big_Data/Test en format 224x224 et à transférer dans le sous repertoire Input_Big_Data/Test1 ¶

In [19]:
Path_Big_Data= r'C:\Users\dgama\Desktop\OPENCLASSROOMS_INFO\REPERTOIRE_P8_BIS\Input_Big_Data\Test1'
from PIL import Image
import glob
lst_imgbig=glob.glob("{0}\**\*.jpg".format(Path_Big_Data),recursive=True)
for i in lst_imgbig:
    folder=i[81:].split(os.sep)[0]
    img=Image.open(i)
    img=img.resize((224,224),Image.LANCZOS)
    if not folder in os.listdir("Input_Big_Data/Test1"):
            os.makedirs("Input_Big_Data/Test1/"+ folder)
    img.save("Input_Big_Data\\Test1\\"+i[81:-4]+".jpg")
    
print(" Job Done")
os.startfile("Input_Big_Data")

 Job Done


## 6_6 Verification de l'effectif des 12455 images reformatées et transférées dans le sous repertoire Input_Big_Data/Test1

In [20]:
Count=0
for dirpath, dirs, files in os.walk("Input_Big_Data/Test1"):
    Count+=len(files)
        
print(f"Nous avons détécté {Count} image de fruits au format 224x224 dans le sous repertoire Input_Big_Data/Test1"  )

Nous avons détécté 0 image de fruits au format 224x224 dans le sous repertoire Input_Big_Data/Test1


# 7 Conversion d'un echantillon de 3110 images de fruits  en format 224x224 et sauvegarde dans le repertoire Test1

In [21]:
from PIL import Image
import glob
        
lst_imgs=[i for i in glob.glob("Test/**/*.jpg")]
for i in lst_imgs:
    folder=i[5:-4].split(os.sep)[0]
    img=Image.open(i)
    img=img.resize((224,224),Image.LANCZOS)
    if not folder in os.listdir("Test1"):
        os.makedirs("Test1/"+ folder)
    img.save("Test1/"+i[5:-4]+".jpg")
print("Job Done")
os.startfile("Test1")

Job Done


## 7_1 Copie des 3110 images reformatées dans le repertoire Test1_Backup

In [22]:
try:
    shutil.copytree('C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Test1','C:\\Users\\dgama\\Desktop\\OPENCLASSROOMS_INFO\\REPERTOIRE_P8_BIS\\Test1_Backup')
except FileExistsError:
   # directory already exists
    pass

## 7_2 Verification de l'effectif 3110 images reformatées et transférées dans le repertoire Test1_Backup

In [23]:
Count=0
for dirpath, dirs, files in os.walk("Test1"):
    Count+=len(files)
        
print(f"Nous avons détécté {Count} image de fruits au format 224x224 dans notre base de donnée"  )

Nous avons détécté 3110 image de fruits au format 224x224 dans notre base de donnée


# 8 Transfert d'un echantillon de 300 fichiers d' images reformatées  vers le repertoire Test2

In [24]:

if not os.path.exists("Test2"):
            os.makedirs("Test2")
    
path_source =cwd+"\\"+"Test1" 
path_target="Test2" 
for path_folder,dirs,files in os.walk(path_source):
    path_fld=path_folder[66:]
    path_file_dest= "Test2"+ path_fld
    if not os.path.exists(path_file_dest):
        os.mkdir(path_file_dest) 
        Count=0
        for file in files:
            if not len(os.lisdir(dest))==0:
                Max=int(len(files)/10)
                Count+=1
                source =path_folder +file
                destination = cwd + path_file_dest
                filename = os.path.basename(source) # In order to overwrite file (if one already exists in destination),
                dest = os.path.join(destination,filename)#we need to specify full path for destination,
                shutil.move(source, dest)
                if Count>Max:
                    break 
            else:
                    break

## 8_1 Controle du transfert des 300 images.

In [25]:
Count=0
for dirpath, dirs, files in os.walk("Test2"):
    Count+=len(files)
        
print(f"\nNous avons un echantillon low_Data de {Count} images à exploiter sur notre environnement de Test\
 avant deploiement Big Data ")        


Nous avons un echantillon low_Data de 295 images à exploiter sur notre environnement de Test avant deploiement Big Data 


# 9 Upload des 12455 images de fruits de notre environnement local vers  le stockage cloud AWS_S3 avec Boto3

### Selon notre étude la librairie python Boto3 est limitée pour le upload des métadonnées  de notre environnement local vers AWS S3 . En effet, l'automatisation du transfert  des métadonnées d'un environnement local vers un environnement Cloud AWS S3 nécessite  de faire usage des technologies de transfert Big_Data adéquats. Nous poursuivrons nos recherches en vue d'identifier les techniques et méthodes les plus adapées au transfert des métadonnées.
### Par conséquent ,le transfert des métadonnées de l'environnement local vers l'environnement de production s'effectuera manuellement à partir des fonctions upload files ou folders de la console  AWS_S3


# 10 Upload d'un echantillon de 295 fichiers images de fruits reformatées  de notre environnement local vers le Storage Cloud cloud AWS_S3 avec Boto3

In [26]:
import boto3


In [27]:
Path_Low_Data= r'C:\Users\dgama\Desktop\OPENCLASSROOMS_INFO\REPERTOIRE_P8_BIS\Test2'


## 10_1 Sécurisation de l'acces_key et secret_acces dans l'environnement des variables utilisateur  de notre serveur windows.

In [28]:
AWS_S3_CREDS = {
    "aws_access_key_id":os.getenv("AWS_ACCESS_KEY"),
    "aws_secret_access_key": os.getenv("AWS_SECRET_ACCESS_KEY")
}

## 10_2 Illustration d'un Script d'upload d'un faible échantillon  de fichiers vers AWS_S3 avec Boto3

In [29]:
s3_client = boto3.client('s3',**AWS_S3_CREDS)

In [30]:
s3_client=boto3.client('s3',**AWS_S3_CREDS)
folderlist=glob.glob("{0}\*".format(Path_Low_Data))
filelist=glob.glob("{0}\**\*.jpg".format(Path_Low_Data),recursive=True)
for folder in folderlist :
    folder_s3 ='"'+ folder[67:] +'"'
    print(folder_s3)
    bucket="myshin"
    for filename in filelist:
        key=folder_s3 +'/'+ filename.split("\\")[-1] 
        reponse=s3_client.upload_file(Filename=filename,Bucket=bucket,Key=key)
print("\nJob Done:All files uploaded to AWS_s3")

"apple_6"
"apple_braeburn_1"
"apple_crimson_snow_1"
"apple_golden_1"
"apple_golden_2"
"apple_golden_3"
"apple_granny_smith_1"
"apple_hit_1"
"apple_pink_lady_1"
"apple_red_1"
"apple_red_2"
"apple_red_3"
"apple_red_delicios_1"
"apple_red_yellow_1"
"apple_rotten_1"
"cabbage_white_1"
"carrot_1"
"cucumber_1"
"cucumber_3"
"eggplant_violet_1"
"pear_1"
"pear_3"
"zucchini_1"
"zucchini_dark_1"

Job Done:All files uploaded to AWS_s3


# 11 Bilan du preprocessing Image de notre projet

## 11_1 Environnement Local:L'Entrepot de notre échantillon  de 295 fichiers images de fruits mis en format 224x224  est est  localisé sur dans le repertoire Test2 du serveur Windows

## 11_2 Environnement Cloud:L'Entrepot des métadonnées ,constitué de 12455 fichiers images de fruits  mis en format 224x224 ,est localisé dans le sous repertoire Input_Big_Data/Test1 du serveur windows 